lib/ai/model/output/streaming.go

```diff
@@ -0,0 +1,65 @@
+package output
+
+import (
+	"errors"
+	"io"
+	"strings"
+
+	"github.com/gravitational/trace"
+	"github.com/sashabaranov/go-openai"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/lib/ai/tokens"
+)
+
+// StreamToDeltas converts an openai.CompletionStream into a channel of strings.
+// This channel can then be consumed manually to search for specific markers,
+// or directly converted into a StreamingMessage with NewStreamingMessage.
+func StreamToDeltas(stream *openai.ChatCompletionStream) chan string {
+	deltas := make(chan string)
+	go func() {
+		defer close(deltas)
+
+		for {
+			response, err := stream.Recv()
+			if errors.Is(err, io.EOF) {
+				return
+			} else if err != nil {
+				log.Tracef("agent encountered an error while streaming: %v", err)
+				return
+			}
+
+			delta := response.Choices[0].Delta.Content
+			deltas <- delta
+		}
+	}()
+	return deltas
+}
+
+// NewStreamingMessage takes a string channel and converts it to
+// a StreamingMessage.
+// If content was already streamed, it must be passed through the alreadyStreamed parameter.
+// If the already streamed content contains a prefix that must be stripped
+// (like a marker to identify the kind of response the model is providing),
+// the prefix can be passed through the prefix parameter. It will be stripped
+// but will still be reflected in the token count.
+func NewStreamingMessage(deltas <-chan string, alreadyStreamed, prefix string) (*StreamingMessage, *tokens.AsynchronousTokenCounter, error) {
```

Comment:
The same here, it sounds more like an internal helper rather than something that can be used outside of this module.